dexcom/time.go

```diff
@@ -1,66 +1,220 @@
+	tm, err := time.Parse(time.RFC3339Nano, parsable)
+	if err != nil {
+		return nil, errors.New("time is not parsable")
+	}
+
+	return &Time{
+		Time:          tm,
+		serialized:    serialized,
+		zoneNotParsed: zoneMatches == nil,
+	}, nil
 }
 
-func (t Time) format() string {
-	return t.Time.Format(TimeFormatMilli)
+func TimeFromRaw(raw time.Time) *Time {
+	return &Time{
+		Time:       raw,
+		serialized: raw.Format(time.RFC3339Nano),
+	}
 }
 
-func TimeFromRaw(raw *time.Time) *Time {
-	if raw == nil {
+func TimeFromTime(tm *Time) *Time {
+	if tm == nil {
 		return nil
 	}
 	return &Time{
-		Time: *raw,
+		Time:          tm.Time,
+		serialized:    tm.serialized,
+		zoneNotParsed: tm.zoneNotParsed,
 	}
 }
 
-func TimeFromString(raw *string) *Time {
-	if raw == nil {
+type Time struct {
+	time.Time
+	serialized    string
+	zoneNotParsed bool // Negated so the default of false is correct if struct manually created
+}
+
+func (t *Time) Raw() *time.Time {
+	if t == nil || t.IsZero() {
 		return nil
 	}
+	return &t.Time
+}
 
-	stringValue := *raw
-	var err error
-	var timeValue time.Time
+func (t *Time) ZoneParsed() bool {
+	return !t.zoneNotParsed
+}
+
+func (t Time) MarshalJSON() ([]byte, error) {
+	if t.serialized != "" {
+		return json.Marshal(t.serialized)
+	} else {
+		return json.Marshal(t.Format(time.RFC3339Nano))
+	}
+}
 
-	format := GetTimeFormat(stringValue)
+var zoneRegexp = regexp.MustCompile(`^(.*)(Z|-\d\d:\d\d)$`)
 
-	timeValue, err = time.Parse(format, stringValue)
+func parseDigits(original string, minimum int, maximum int) (*int, string, error) {
+	digits, remaining, err := parseCharacters(original, "1234567890", minimum, maximum)
 	if err != nil {
-		return nil
+		return nil, original, err
+	}
+	number, err := strconv.ParseInt(*digits, 10, 64)
+	if err != nil {
+		return nil, original, errors.Wrap(err, "string is not a valid number")
 	}
+	return pointer.FromInt(int(number)), remaining, nil
+}
 
-	return &Time{
-		Time: timeValue,
+func parseCharacter(original string, character string) (string, error) {
+	_, remaining, err := parseCharacters(original, character, 1, 1)
+	return remaining, err
+}
+
+func parseCharacters(original string, characters string, minimum int, maximum int) (*string, string, error) {
+	if minimum < 0 {
+		return nil, original, errors.New("minimum is less than zero")
+	} else if maximum < minimum {
+		return nil, original, errors.New("maximum is less than minimum")
+	}
+
+	if maximum == 0 {
+		return pointer.FromString(""), original, nil
+	}
+
+	var ruins []rune
+	for _, ruin := range original {
+		if !strings.ContainsRune(characters, ruin) {
+			break
+		}
+		ruins = append(ruins, ruin)
+		if len(ruins) >= maximum {
+			break
+		}
+	}
+	if len(ruins) < minimum {
+		return nil, original, errors.New("string does not contain minimum number of characters")
+	}
+	return pointer.FromString(string(ruins)), original[len(ruins):], nil
+}
+
+func intOrDefault(value *int, defowlt int) int {
```

Comment:
Ha! :) No, that would be a made up substitution. I didn't want to use `default` because it is a keyword, so this is just an alternate spelling that sounds about the same. Just like `bites` (for `bytes`, which is package), `lgr` (for `logger`, which is a package), etc...